[user-triton] handle inline_asm_case

sijiac opened the pull request:
Summary: We currently failed the mutation analysis for all inline_asm ops. In this diff, we handle the case when "is_pure" is set to True since it indicates the operation doesn't mutate the input value

Test Plan:
../buck-out/v2/gen/fbcode/854b9ed00d28c5c5/caffe2/test/inductor/__triton_kernels__/triton_kernels.par --r test_mutations_inline_asm_kernel

```
test_mutations_inline_asm_kernel_is_pure_true (caffe2.test.inductor.test_triton_kernels.MutationTests) ... W0226 18:10:34.261000 1906801 /data/users/sijiac/fbsource/fbcode/caffe2/torch/_higher_order_ops/triton_kernel_wrap.py:656] TTIR mutation analysis: Skipping pure tt.elementwise_inline_asm op (is_pure=True)
ok

----------------------------------------------------------------------
Ran 2 tests in 0.706s

OK
```

Differential Revision: D69878591




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

sijiac commented:
@pytorchbot label "topic: not user facing"

zou3519 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -640,7 +649,12 @@ def analyze_kernel_mutations(
     for op_list in ops.values():
         for op in op_list:
+            # If we encounter an operation with effects that cannot be reliably analyzed
+            # (e.g. `tt.elementwise_inline_asm`), we assume it does not mutate any input parameters.
```
We assume it does not mutate any input parameters only if it has `is_pure=True` right?

sijiac replied:
yes, it's more like if is_pure = True, we assume not mutate of input args

zou3519 commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3127,9 +3127,13 @@ def branch_with_multiple_yield_args(
+            inline_asm_kernel_is_pure_true,
+            {"X": t, "Y": t, "Z": t, "n": 4, "BLOCK": 4},
+            ["Z"],
+        ],
```
Do you have a test that shows we do not failure the mutation check on  inline_asm with `is_pure`?

sijiac replied:
sry, i didn't catch it. You mean the test case of `is_pure` is set to `False`?

zou3519 replied:
Did this test case previously fail (raise a runtime error), or was it a perf issue?

If the test case previously failed without your PR, then I'm happy.

If the test case previously was just a perf issue (and not a correctness issue), then we may need a stronger test to prevent this from regressing.

sijiac replied:
It's not the perf issue. It's a kernel with inline_asm would fail to compiling, reported in the user-group. After rounds of debugging, it pointed to the feature missing of handling inline_asm mutation analysis

zou3519 replied:
sounds good, thanks for clarifying

zou3519 reviewed (commented):
code looks reasonable, this needs a test

zou3519 commented:
btw lintrunner is failing, you'll want to fix that before merging

sijiac committed:
[user-triton] handle inline_asm_case (#148043)

Summary:

We currently failed the mutation analysis for all inline_asm ops. In this diff, we handle the case when "is_pure" is set to True since it indicates the operation doesn't mutate the input value

Test Plan:
../buck-out/v2/gen/fbcode/854b9ed00d28c5c5/caffe2/test/inductor/__triton_kernels__/triton_kernels.par --r test_mutations_inline_asm_kernel

```
test_mutations_inline_asm_kernel_is_pure_true (caffe2.test.inductor.test_triton_kernels.MutationTests) ... W0226 18:10:34.261000 1906801 /data/users/sijiac/fbsource/fbcode/caffe2/torch/_higher_order_ops/triton_kernel_wrap.py:656] TTIR mutation analysis: Skipping pure tt.elementwise_inline_asm op (is_pure=True)
ok

----------------------------------------------------------------------
Ran 2 tests in 0.706s

OK
```

Reviewed By: oulgen, zou3519

Differential Revision: D69878591

sijiac commented:
@pytorchbot merge
